src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -5433,87 +5433,6 @@ void C2_MacroAssembler::vector_count_trailing_zeros_avx(BasicType bt, XMMRegiste
   vpsub(bt, dst, xtmp1, dst, vec_enc);
 }
 
-void C2_MacroAssembler::udivI(Register rax, Register divisor, Register rdx) {
-  Label done;
-  Label neg_divisor_fastpath;
-  cmpl(divisor, 0);
-  jccb(Assembler::less, neg_divisor_fastpath);
-  xorl(rdx, rdx);
-  divl(divisor);
-  jmpb(done);
-  bind(neg_divisor_fastpath);
-  // Fastpath for divisor < 0:
-  // quotient = (dividend & ~(dividend - divisor)) >>> (Integer.SIZE - 1)
-  // See Hacker's Delight (2nd ed), section 9.3 which is implemented in java.lang.Long.divideUnsigned()
```

Comment:
Where is ideal implementation of this optimization (you changed `.ad` to handle only positive cases)?
I see you mixed positive and negative divisor values in JMH tests. Can you do separately?

Reply:
An ideal rule is added to transform the division if the divisor is known to be negative, I don't think a runtime branch is beneficial, and users of the method can always do it from Java if the data is expected to have a significant rate of negative divisor. WDYT?

Reply:
I am fine with ideal transformation instead of runtime check in code. The problem is I can't find this new ideal transformation in your changes. I don't see check for **not constant** negative divisor.
I would leave **not constant devisor** case alone. It could be problematic to introduce control flow into ideal graph for these nodes.